src/main/java/uk/ac/ebi/subs/fileupload/services/DefaultEventHandlerService.java

```diff
@@ -55,4 +67,35 @@ public DefaultEventHandlerService(ValidationService validationService, FileRepos
 
         return new ResponseEntity<>(HttpStatus.OK);
     }
+
+    @Override
+    public void moveFile(File file) throws IOException {
+        String sourceFileName = file.getTusId();
+
+        String targetPath = generateFolderName(file.getTusId());
+
+        createTargetFolder(sourcePath, targetBasePath, targetPath);
+
+        String targetFilename = file.getFilename();
+
+        String fullTargetPath = assembleFullTargetPath(sourcePath, targetBasePath, targetPath, targetFilename);
+
+        String fullSourcePath = assembleFullSourcePath(sourcePath, sourceFileName);
+
+        Files.move(Paths.get(fullSourcePath), Paths.get(fullTargetPath), StandardCopyOption.ATOMIC_MOVE);
+    }
+
+    private String generateFolderName(String tusId) {
```

Comment:
Do we need to generate a random number here? The submission ID is already random. You could take the first x chars, then the second x chars, and then the whole submission ID for folder names. e.g. if the submission ID is abcd-efgh-3490234234, the path could be /ab/cd/abcd-efgh-3490234234